src/raft/raft.rs

```diff
@@ -798,11 +856,92 @@ impl<T: Storage> Raft<T> {
                 return Ok(());
             }
         }
-        match self.state {
-            StateRole::Candidate => self.step_candidate(m),
-            StateRole::Follower => self.step_follower(m),
-            StateRole::Leader => self.step_leader(m),
+
+
+        match m.get_msg_type() {
+            MessageType::MsgHup => {
+                if self.state != StateRole::Leader {
+                    let ents = self.raft_log
+                        .slice(self.raft_log.applied + 1,
+                               self.raft_log.committed + 1,
+                               raft_log::NO_LIMIT)
+                        .expect("unexpected error getting unapplied entries");
+                    let n = self.num_pending_conf(&ents);
+                    if n != 0 && self.raft_log.committed > self.raft_log.applied {
+                        warn!("{} cannot campaign at term {} since there are still {} pending \
+                               configuration changes to apply",
+                              self.tag,
+                              self.term,
+                              n);
+                        return Ok(());
+                    }
+                    info!("{} is starting a new election at term {}",
+                          self.tag,
+                          self.term);
+                    if self.pre_vote {
+                        self.campaign(CAMPAIGN_PRE_ELECTION);
+                    } else {
+                        self.campaign(CAMPAIGN_ELECTION);
+                    }
+                } else {
+                    debug!("{} ignoring MsgHup because already leader", self.tag);
+                }
+            }
+            MessageType::MsgRequestVote |
+            MessageType::MsgRequestPreVote => {
+                // The m.get_term() > self.term clause is for MsgRequestPreVote. For MsgRequestVote
+                // m.get_term() should always equal self.term
+                if (self.vote == INVALID_ID || m.get_term() > self.term ||
+                    self.vote == m.get_from()) &&
+                   self.raft_log.is_up_to_date(m.get_index(), m.get_log_term()) {
+                    info!("{} [logterm: {}, index: {}, vote: {}] cast {:?} for {} [logterm: {}, \
```

Comment:
The function is too long, it's more readable to keep these two line consuming apart.